framework/src/utils/RankThreeTensor.C

```diff
@@ -0,0 +1,307 @@
+{
+  for (unsigned int i = 0; i < N; ++i)
+    for (unsigned int j = 0; j < N; ++j)
+      for (unsigned int k = 0; k < N; ++k)
+        _vals[i][j][k] -= a(i, j, k);
+
+  return *this;
+}
+
+RankThreeTensor
+RankThreeTensor::operator-(const RankThreeTensor & b) const
+{
+  RankThreeTensor result;
+  const RankThreeTensor & a = *this;
+
+  for (unsigned int i = 0; i < N; ++i)
+    for (unsigned int j = 0; j < N; ++j)
+      for (unsigned int k = 0; k < N; ++k)
+        result(i, j, k) = a(i, j, k) - b(i, j, k);
+
+  return result;
+}
+
+RankThreeTensor
+RankThreeTensor::operator-() const
+{
+  RankThreeTensor result;
+  const RankThreeTensor & a = *this;
+
+  for (unsigned int i = 0; i < N; ++i)
+    for (unsigned int j = 0; j < N; ++j)
+      for (unsigned int k = 0; k < N; ++k)
+        result(i, j, k) = -a(i, j, k);
+
+  return result;
+}
+
+Real
+RankThreeTensor::L2norm() const
+{
+  Real l2 = 0;
+  const RankThreeTensor & a = *this;
+
+  for (unsigned int i = 0; i < N; ++i)
+    for (unsigned int j = 0; j < N; ++j)
+      for (unsigned int k = 0; k < N; ++k)
+        l2 += Utility::pow<2>(a(i, j, k));
+
+  return std::sqrt(l2);
+}
+
+void
+RankThreeTensor::fillFromInputVector(const std::vector<Real> & input, FillMethod fill_method)
+{
+  zero();
+
+  switch (fill_method)
+  {
+    case general:
+      fillGeneralFromInputVector(input);
+      break;
+    default:
+      mooseError("fillFromInputVector called with unknown fill_method of ", fill_method);
+  }
+}
+
+void
+RankThreeTensor::rotate(const RealTensorValue & R)
+{
+  RankThreeTensor old = *this;
+
+  for (unsigned int i = 0; i < N; ++i)
+    for (unsigned int j = 0; j < N; ++j)
+      for (unsigned int k = 0; k < N; ++k)
+      {
+        Real sum = 0.0;
+        for (unsigned int m = 0; m < N; ++m)
+          for (unsigned int n = 0; n < N; ++n)
+            for (unsigned int o = 0; o < N; ++o)
+              sum += R(i, m) * R(j, n) * R(k, o) * old(m, n, o);
+
+        _vals[i][j][k] = sum;
+      }
+}
+
+void
+RankThreeTensor::rotate(const RankTwoTensor & R)
+{
+  RankThreeTensor old = *this;
+
+  for (unsigned int i = 0; i < N; ++i)
+    for (unsigned int j = 0; j < N; ++j)
+      for (unsigned int k = 0; k < N; ++k)
+      {
+        Real sum = 0.0;
+        for (unsigned int m = 0; m < N; ++m)
+          for (unsigned int n = 0; n < N; ++n)
+            for (unsigned int o = 0; o < N; ++o)
+              sum += R(i, m) * R(j, n) * R(k, o) * old(m, n, o);
+
+        _vals[i][j][k] = sum;
+      }
+}
+
+void
+RankThreeTensor::fillGeneralFromInputVector(const std::vector<Real> & input)
+{
+  if (input.size() != 18)
+    mooseError("To use fillGeneralFromInputVector, your input must have size 18. Yours has size ",
+               input.size());
+
+  int ind;
+  for (unsigned int i = 0; i < N; ++i)
+    for (unsigned int j = 0; j < N; ++j)
+      for (unsigned int k = 0; k < N; ++k)
+      {
+        ind = i * N * N + j * N + k;
```

Comment:
2*3*3+2*3+2 = 26 that's out of bounds. Of course a 3x3x3 tensor has 27 entries. I thought I was losing my mind...